### ISSUE TITLE
Sound cuts out frequently, repeatedly Fedora 39, Gnome 45

### ISSUE THREAD

**Issue by douglasthiel** [maniacx/Airpods-Battery-Monitor#13]
Environment: [Macmini7,1](https://support.apple.com/kb/SP710?locale=en_US)
Airpods 3rd Gen
Fedora 39
Gnome 45
Firefox 123.0
Bluetooth 4.0 wireless technology
> lsusb | grep bluetooth -i
> Bus 001 Device 003: ID 0a5c:4500 Broadcom Corp. BCM2046B1 USB 2.0 Hub (part of BCM2046 Bluetooth)
> Bus 001 Device 007: ID 05ac:8289 Apple, Inc. Bluetooth Host Controller

Repeatable test sequence:
1. Turn on extension.
2. Listen to YouTube or Rhythmbox
3. Observe cracks and silences
4. Turn off extension
5. Observe proper play

**Comment by maniacx** [comment]
Could be the start and stop discovery  might be cause issue with streaming data.

Let do some test with bluetoothctl tool
in bluetoothctl commands are
scan on: to start discovery normal mode
scan le: to start discovery using filter for ble only (This seem to get faster response whihc the extension is using)
scan off: to stop discovery

Do both test with youtube playback and extension disabled. start discovery and stop discovery after 1 minute. when start discovery is runnng, observe if it affect streaming

Test 1 start discovery with le filter. In terminal type
`bluetoothctl`
`scan le`
 check if it affect streaming for a minute
`scan off`


Test 2 start discovery. In terminal type
`bluetoothctl`
`scan on`
 check if it affect streaming for a minute
`scan off`


Let me know if test1 and test2 interfere with streaming, or if one is better then other.

**Comment by douglasthiel** [comment]
Both Test 1 and Test 2 significantly interfere with YouTube audio

**Comment by maniacx** [comment]
Oh damn.
Ok one more. test With everything closed and extension disabled.
Run Youtube.
Open Gnome Settings > Bluetooth section and just leave it open.
Check if leaving this windows open, you have  any interference.

![Screenshot from 2024-02-28 16-53-16](https://github.com/maniacx/Airpod-Battery-Monitor/assets/4295343/0cc7d2a4-ce61-48be-b3b6-da43e1093911)


Explaination. Opening Gnome setting > Bluetooth also starts discovery.
Just want to check if this too interferes with streaming.

**Comment by douglasthiel** [comment]
Listening to YouTube with the Airpods, Opening Gnome Bluetooth Settings has the same result interfering with playback.
Closing the Gnome Bluetooth Settings restores normal playback.

What does it mean "With everything closed" ??  All apps closed?
Anyway, I retested after closing all apps except Firefox 123.0 and the Gnome Settings app with the same result.

**Comment by maniacx** [comment]
> What does it mean "With everything closed" ?? All apps closed?

No just the bluetoothctl window or any thing that start discovery.

The extension depends on discovery to get battery info from airpods. For now, I cannot think of anything else.

One thing you can do is experement with different codec configurations if there are multiple configuration available and test with discovery on in bluetoothctl.

![Screenshot from 2024-02-28 17-23-38](https://github.com/maniacx/Airpod-Battery-Monitor/assets/4295343/3f05f615-f74e-4650-9615-c5de7cda64f0)

Let me know if something improves

**Comment by douglasthiel** [comment]
All tests were performed with the following codec:
High Fidelity Playback (A2DP Sink, codec SBC-XQ)

**Comment by maniacx** [comment]
Do you have only 1 configuration available?

**Comment by douglasthiel** [comment]
Testing with the other 2 codecs on my system yields the same result. They are:
High Fidelity Playback (A2DP Sink, codec AAC)
High Fidelity Playback (A2DP Sink, codec SBC)

**Comment by Toxblh** [comment]
@douglasthiel could you test to change ControllerMode in `/etc/bluetooth/main.conf` to `ControllerMode = bredr`
and test again?

**Comment by douglasthiel** [comment]
ControllerMode in /etc/bluetooth/main.conf was originally like this:
#ControllerMode

Changed to:
ControllerMode = bredr
rebooted
all Fedora updates applied (yet again)

Retested with the same result.

**Comment by maniacx** [comment]
Are you able to see Airpods Battery Level with `ControllerMode = bredr`?

Also could you post result of 
```
cat /etc/bluetooth/main.conf
```
```
bluetoothctl --version
```
```
pipewire --version
```

**Comment by maniacx** [comment]
Are you using asahi linux kernel? Is your CPU intel or M-series?

**Comment by douglasthiel** [comment]
[main.conf.txt](https://github.com/maniacx/Airpod-Battery-Monitor/files/14455753/main.conf.txt)

**Comment by douglasthiel** [comment]
> bluetoothctl --version
> bluetoothctl: 5.72

**Comment by douglasthiel** [comment]
> pipewire --version
> pipewire
> Compiled with libpipewire 1.0.3
> Linked with libpipewire 1.0.3

**Comment by douglasthiel** [comment]
> Are you using asahi linux kernel? Is your CPU intel or M-series?

The model for Apple Silicon macmini starts at [macmini9,1](https://everymac.com/systems/apple/mac_mini/specs/mac-mini-m1-8-core-2020-specs.html)
So, my [macmini7,1](https://support.apple.com/kb/SP710?locale=en_US) 2014 is running Intel

**Comment by douglasthiel** [comment]
> Are you able to see Airpods Battery Level with ControllerMode = bredr?   

Yes. The Airpods Battery Level is in the process of "scanning", seemingly forever.
1. I took my Airpods out of my ears and put them back in the case. 
2. I took them out again and Airpods Battery Level is still in the "scanning" state.

**Comment by maniacx** [comment]
Every look good and updated. I am not sure what could be the issue. But there are few things we could try.
But before that switch back to `ControllerMode = dual` and confirm the extension show battery level and music distorts.


1) Try turning off wifi. and check if music distorts when discovery is started.

2)  Buffer size is dynamic. Try setting a fixed or increasing buffer size by running command .

In terminal
```
pw-metadata -n settings 0 clock.force-quantum 2048
```
To reset back to dyanamic buffer
```
pw-metadata -n settings 0 clock.force-quantum 0
```

3) Also could try these settings 

In terminal
```
PULSE_LATENCY_MSEC=100
```

Do not try all steps together. Try 1 then 2 and finally 3.
Note: that point 2 and 3 are not permanent settings. These settings will be lost on reboot. So it is pretty safe to use.

If none make any difference, let raise an issue with bluez.  Let me know how it goes.

**Comment by Toxblh** [comment]
I think here problem with broardcom bt. I'll test with my macbook pro 2013 later. 
With intel bt all good in regular PC

**Comment by Toxblh** [comment]
Confirmed. On MacbookPro 11,1 get the same super annooying problem with sound under listen advertisments. 
I use close source Broadcom drivers for wifi/bt combo. 

Will see.

**Comment by douglasthiel** [comment]
> Every look good and updated. I am not sure what could be the issue. But there are  few things we could try.
> But before that switch back to ControllerMode = dual and confirm the extension show battery level and music distorts.

check, same issue as always

>    Try turning off wifi. and check if music distorts when discovery is started.

macmini is hardwired to my router.

I tried buffer sizes of 2048, 4096, 8192 with no joy.

> pw-metadata -n settings 0 clock.force-quantum 0

no joy

>PULSE_LATENCY_MSEC=100

also no joy

Note: I am not rebooting btwn the different tries in this comment.
